### DISCUSSION TITLE
ASPLOS'24 paper on EVT (Epilogue Visitor Tree)

### DISCUSSION THREAD

**Discussion by hwu36** [NVIDIA/cutlass#1510] [General]
@apuaaChen is going to present his CUTLASS [paper](https://dl.acm.org/doi/10.1145/3620666.3651369) - EVT: Accelerating Deep Learning Training with Epilogue Visitor Tree in ASPLOS'24 on May 1.

EVT is a framework to fuse almost any combination in the epilogue (elementwise, reduction, broadcast, etc.). If you attend ASPLOS and use CUTLASS, highly recommend to talk with him.

**Comment by ginowu** [comment]
On page 13 of the slides(CUTLASS: Python API, Enhancements, and CUTLASS 3.0 Preview) in GTC22, it lists some advanced fusion patterns that supported by cutlass-python, do you recommend using EVT to realize these patterns? If so, could you elaborate on the details to achieve it?


<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>

<meta http-equiv="Content-Style-Type" content="text/css">
<meta name="Generator" content="Cocoa HTML Writer">
<meta name="CocoaVersion" content="2487.5">
</head>
<body>

Mainloop | Elementwise compute | Conversion, scaling, masking
-- | -- | --
Reduction of A or B | GEMM + bias grad
Broadcast | Load vector and broadcast over channels
Epilogue | Multiplicity of tensor operands | GEMM + GELU with Aux tensor; GEMM + RELU with output bitmask GEMM + dRELUloading bitmask
Multiplicity of vector operands | Alpha and beta scaling as vector
Broadcast over column | Bias add
Reduce over column | GEMM + bias Grad Layer norm
Reduce over row | GEMM + Softmax
Elementwise compute | Arithmetic, conversion, activation functions


</body>
</html>
